### ISSUE TITLE
Android: VNCConn Refactoring

### ISSUE THREAD

**Issue by gujjwal00** [bk138/multivnc#46]
VNCConn refactoring Proposal (as per #15 ):

**Extract progress dialog & fatal error alert**
- Move ProgressDialog and error dialogs to `VncCanvasActivity`.
- `VNCConn` will expose an  observer interface which will be implemented by `VncCanvasActivity`.
- `VNCConn` will publish updates through the observer.

This will allow the activity to attach/detach to `VNCConn` without affecting its operation. It also simplifies `VNCConn` and makes it UI agnostic.

**Move Credential dialog**
- Move credential dialog from `VncCanvas` to `VncCanvasActivity` .
- `VNCConn` will notify the activity through observer.
- Activity will display the dialog.

In future when we implement rotation, it will allow the activity to save the credential request in its ViewModel to handle restarts. It will also reduces the direct dependency on `VncCanvas`. 

**Extract FrameBuffer**
- Create a new 'FrameBuffer' class.
- Move bitmap data from `VNCConn` to `FrameBuffer`
- Move pointer position from `VncCanvas` to `FrameBuffor`
- Attach `FrameBuffer` with `VncCanvas` and pass a reference to `VNCConn`

Now we will be able to generate a new `FrameBuffer` whenever required. It also enables `VNCConn` to function without any fame buffer (time between activity restarts). See figure below.

-------------------------------------------------------------------------------------------------------------------------------------
After this refactoring we should be able to put `VNCConn` in a ViewModel which will persist across activity restarts. Now we can implement whichever strategy (Restarts vs No restart) we choose.

**For Restarts:**
![image](https://user-images.githubusercontent.com/27951674/74340143-67006380-4dcb-11ea-9c93-f11826eecc86.png)
We detach from `VNCConn` inside `onStop()`  and attach again in `onStart()` of new activity

**For No restart**
![image](https://user-images.githubusercontent.com/27951674/74340535-12111d00-4dcc-11ea-9b9f-641339aa7cd0.png)
We can simply generate a new `FrameBuffer` for the resized `VncCanvas` and pass it to `VNCConn`

*Data flow*
![image](https://user-images.githubusercontent.com/27951674/74340771-906dbf00-4dcc-11ea-8531-7333fbb3a006.png)

----------------------------------------------------------------------------------------------------------------------------------
**Future/Optional**
These are few ideas we can try after refactoring is complete.
- Now that all dialogs/alerts are handled by `VncCanvasActivity`, we can unify them into one custom dialog. It will allow us to replace `ProgressDialog` which is deprecated and it will integrate better with ViewModel.
- Rename `ServerToClientThread` to `ReceiverThread` & `ClientToServerThread` to `SenderThread`.

Let me know if you have any suggestion or something is wrong.
Thanks.

**Comment by gujjwal00** [comment]
Iteration 2:
--------------
We can make make `FrameBuffer` independent of `VncCanvas` with few simple tweaks:
- `VncCanvas` is used by bitmap-data classes .
- Access to `VncCanvas` inside these classes can be replaced with method arguments.

Now `FrameBuffer` can be moved to [ViewModel](https://developer.android.com/topic/libraries/architecture/viewmodel.html).  

**For Restarts:**
![image](https://user-images.githubusercontent.com/27951674/74419268-a6808b80-4e6f-11ea-8f56-7c7cec556ac9.png)

**For No-restart:**
![image](https://user-images.githubusercontent.com/27951674/74419225-949ee880-4e6f-11ea-89c2-64f27f568c17.png)


Now  `FrameBuffer` needs to be created only once.

**Comment by gujjwal00** [comment]
Can you please answer few questions:
1. `LargeBitmapData` is not used anymore (conditional branch for its creation is never executed). Even if I force its use, I can't get past the "Loading first frame" dialog. Is this intentional or a bug?

2. What is 'local cursor' ? I see few references to `setUseLocalCursor`/`getUseLocalCursor`.

**Comment by gujjwal00** [comment]
Closing this in favor of #56 . It will need some rework.
